lte/gateway/c/oai/tasks/mme_app/mme_app_detach.c

```diff
@@ -154,8 +154,7 @@ void mme_app_handle_detach_req(const mme_ue_s1ap_id_t ue_id) {
       "Handle Detach Req at MME app for ue-id: " MME_UE_S1AP_ID_FMT
       " with mme_s11_teid " TEID_FMT "\n",
       ue_id, ue_context_p->mme_teid_s11);
-  if ((!ue_context_p->mme_teid_s11) &&
```

Comment:
why is the check for `mme_teid_s11` no longer needed? @pruthvihebbani I remember that you introduced this to fix another issue long time back?

Reply:
From the logs below is my observation; 
1. On mme restart, for non-registered UEs, mme would implicitly detach the UEs. 
2. Referring to mme_imsi_leftoverstate.txt, contexts are left for imsi, 1011234560213 and Within this UE's context, I couldn't notice the field, nb_active_pdn_contexts, but in state conversion logic this field is written to DB. 
And referring to mme_nas_state.txt, wrt to this UE, there are entries left in imsi and s11_teid hash table; below is the snapshot
mme_ue_contexts {
  imsi_ue_id_htbl {
    key: 1011234560213
    value: 214
  }
  imsi_ue_id_htbl {
    key: 1011234560291
    value: 1679
  }
  tun11_ue_id_htbl {
    key: 214
    value: 214
  }
3. Referring to mme.log for this UE, below is logs captured for this UE

491 000644 Sun May 23 18:06:11 2021 7FED3F8EF700 INFO  MME-AP tasks/mme_app/mme_app_context.c :2344   [1011234560213] Initiate implicit detach for ue_id0x000000D6
    492 000645 Sun May 23 18:06:11 2021 7FED3F8EF700 INFO  NAS-EM tasks/nas/emm/sap/emm_cn.c      :1136    EMMCN-SAP - Received primitive EMMCN_IMPLICIT_DETACH_UE (407)
    493 000646 Sun May 23 18:06:11 2021 7FED3F8EF700 INFO  NAS-EM tasks/nas/emm/Detach.c          :0265    EMM-PROC  - SGS Detach type = NW-INITIATED-EPS (3) requested (ue_id=0x000000D6)
    494 000648 Sun May 23 18:06:11 2021 7FED3F8EF700 INFO  NAS-EM tasks/nas/emm/Detach.c          :0332    EMM-PROC  - Detach type = EPS (0) requested (ue_id=0x000000D6)
    495 000647 Sun May 23 18:06:11 2021 7FED3D0EA700 INFO  SPGW-A tasks/sgw/sgw_handlers.c        :1057   [1011234560260] Handle delete session request for sgw_s11_teid 0x0
    496 000649 Sun May 23 18:06:11 2021 7FED3F8EF700 INFO  NAS-EM tasks/nas/emm/sap/emm_fsm.c     :0238    EMM-FSM   - Received event DETACH_REQ (9) in state EMM-DEREGISTERED
    497 000650 Sun May 23 18:06:11 2021 7FED3F8EF700 ERROR NAS-EM tasks/nas/emm/sap/EmmDeregistere:0193    EMM-FSM state EMM_DEREGISTERED - Primitive _EMMREG_DETACH_REQ is not valid
    498 000651 Sun May 23 18:06:11 2021 7FED3F8EF700 INFO  MME-AP tasks/mme_app/mme_app_detach.c  :0152    Handle Detach Req at MME app for ue-id: 0x000000D6 with mme_s11_teid 0xD6

Later it was not processed further. Because for this UE, mme_s11_teid value is 214 and nb_active_pdn_contexts is zero, due to which, mme is not able to initiate delete session requests towards SPGW. 
Considering normal attach and detach flow, on reception of Delete Session Response, mme decreases the value of nb_active_pdn_contexts. Long back we were setting the value of mme_s11_teid to zero, so we needed this check. Since mme_s11_teid maps to UE, not for PDN. On reception of Delete Session Response, we removed the code that resets the value of mme_s11_teid from here.

Conclusion: On reception of Delete Session Response, mme decreases the value of nb_active_pdn_contexts, so while handling detach procedure, mme looks for nb_active_pdn_contexts, if it’s non-zero and if pdn_context is active, mme shall send Delete Session Request.

@pruthvihebbani  Could you confirm whether changes are correct

Reply:
@ulaskozat This is not my fix. I had made changes in mme_app/mme_app_bearer.c file. We were setting ue_context_p->mme_teid_s11 to 0 even if removal of mme_teid_s11 from the hashtable was not successful. https://github.com/magma/magma/pull/4141/files